File: examples/python/oracle-price/OraclePriceFeed.runar.py
from runar import (
    SmartContract, PubKey, Sig, ByteString, RabinSig, RabinPubKey, Bigint,
    public, assert_, check_sig, verify_rabin_sig, num2bin,
)

class OraclePriceFeed(SmartContract):
    oracle_pub_key: RabinPubKey
    receiver: PubKey

    def __init__(self, oracle_pub_key: RabinPubKey, receiver: PubKey):
        super().__init__(oracle_pub_key, receiver)
        self.oracle_pub_key = oracle_pub_key
        self.receiver = receiver

    @public
    def settle(self, price: Bigint, rabin_sig: RabinSig, padding: ByteString, sig: Sig):
        msg = num2bin(price, 8)
        assert_(verify_rabin_sig(msg, rabin_sig, padding, self.oracle_pub_key))
        assert_(price > 50000)
        assert_(check_sig(sig, self.receiver))

File: examples/ruby/oracle-price/OraclePriceFeed.runar.rb
require 'runar'

class OraclePriceFeed < Runar::SmartContract
  prop :oracle_pub_key, RabinPubKey
  prop :receiver, PubKey

  def initialize(oracle_pub_key, receiver)
    super(oracle_pub_key, receiver)
    @oracle_pub_key = oracle_pub_key
    @receiver = receiver
  end

  runar_public price: Bigint, rabin_sig: RabinSig, padding: ByteString, sig: Sig
  def settle(price, rabin_sig, padding, sig)
    msg = num2bin(price, 8)
    assert verify_rabin_sig(msg, rabin_sig, padding, @oracle_pub_key)
    assert price > 50_000
    assert check_sig(sig, @receiver)
  end
end
